### PR TITLE
Add NativeOverlapped overloads

### LINKED ISSUE #419: Review and Improve Overlapped IO Policies
Doing overlapped IO with the transition between managed and unmanaged code is significantly difficult. While the overlapped structure is recreated, and several methods have been provided, without additional work the process for actually using them is unclear and extremely error prone. 

The .NET framework and .NET core both contain facilities that make this process easier. Examples are `Overlapped`/`NativeOverlapped` and `ThreadPoolBoundHandle`. Since NativeOverlapped shares the same layout as the provided OVERLAPPED struct, using a pointer cast works, but seems to me a bit messy. 

I'd like to start a discussion about how this library should handle this particular area.

### PR DESCRIPTION
**Pull request by qmfrederik**
In line with #419, and also based on my own experience, add overloads which use the `System.Threading.NativeOverlapped` struct instead of `Kernel32+OVERLAPPED`. .NET has built-in support for overlapped I/O, so let's use that instead of rolling our own.

In line with the guidance in #487, obsolete these types/overloads instead of removing them.

Closes #419

### CODE REVIEW

**Comment by qmfrederik** [comment]
Hmm, having overloads with both `OVERLAPPED*` and `NativeOverlapped*` structs generates conflicts (the code generator would generate two identical overloads with `IntPtr` parameters), so I removed `OVERLAPPED*` for now. Happy to revisit if needed.

**Comment by AArnott** [comment]
This is a binary breaking change. AFAIK we haven't introduced any lately, so I hesitate to take one now unless we're prepared to do a bunch of others as well to minimize the number of releases with breaking changes over time. #288 tracks many ideas for breaking changes.
We could perhaps start a branch with changes like these and merge them all at once, although I'm concerned about merge conflicts too, so I prefer to keep a list of ideas for changes and then we can work on them all at once and merge more quickly.

**Comment by qmfrederik** [comment]
@AArnott Alternatively, we can add this as a new overload in a `.Helper.cs` file; but at the moment the code generators would generate an `IntPtr`-based overload, which would cause conflicts.

**Comment by AArnott** [comment]
What if we just add conversion operators to our struct so it can be cast to the other one?

**Review by AArnott — CHANGES_REQUESTED** [review]
I've heard from the .NET interop folks that each p/invoke function has a fixed cost and they were very happy to hear that we had just one per Win32 function and all our overloads were just thunks into that same one. Rather than creating new p/invokes for the same function, can you add methods to the helper file that simply re-cast the pointer from `NativeOverlapped*` to `OVERLAPPED*`? That way we have more performant, less memory hungry code.

File: src/Kernel32/Kernel32.Helpers.cs [hunk 1]
```diff
@@ -256,6 +257,293 @@ public static ArraySegment<byte> ReadFile(SafeObjectHandle hFile, int nNumberOfB
...
+        ///     <para>
+        ///         If the function fails, the return value is 0 (zero). To get extended error information, call the
+        ///         <see cref="GetLastError" /> function.
+        ///     </para>
+        ///     <para>
+        ///         If this function cannot find a request to cancel, the return value is 0 (zero), and
+        ///         <see cref="GetLastError" />
+        ///         returns <see cref="Win32ErrorCode.ERROR_NOT_FOUND" />.
+        ///     </para>
+        /// </returns>
```
Comment (AArnott) [inline src/Kernel32/Kernel32.Helpers.cs]: Do you know you don't have to copy this from the other overload? 
You can replace it all with this:
```cs
/// <inheritdoc cref="CancelIoEx(SafeObjectHandle, OVERLAPPED*)" />
```

Then, if you hover over your new method's name or its parameters, you'll see that the docs all magically show up.

**Review by AArnott — CHANGES_REQUESTED** [review]
Looks perfect. But we could trim down the xml doc comment weight of it significantly. See my tip.

**Review by AArnott — APPROVED** [review]

**Pull request merged**